common/array.h

```diff
@@ -219,11 +219,9 @@ class Array {
 	/** Remove an element at the given position from the array and return the value of that element. */
 	T remove_at(size_type idx) {
 		assert(idx < _size);
-		T tmp = _storage[idx];
-		copy(_storage + idx + 1, _storage + _size, _storage + idx);
+		T tmp = Common::move(_storage[idx]);
+		uninitialized_move(_storage + idx + 1, _storage + _size, _storage + idx);
```

Comment:
You need to call `_storage[idx].~T()` before `uninitialize_move` here, or just call `erase` which already does it correctly.

Reply:
Wouldn't it be enough to use `Common::move(_storage[idx]);` as the previous line does? If not, that could pose a problem with the implementation of `uninitialized_move` itself.

Reply:
The second option looks preferable, but I'm not sure how to do that.

Reply:
In STL the way it is done is that there are `move` and `move_backwards` functions that work the same as `copy` and `copy_backwards` except it calls single-parameter `move` on the source operands.

Then for most of these cases, you just call `move`/`move_backwards` instead of `copy`/`copy_backwards` and keep the destructor calls.

Reply:
OK, I think I've done this correctly.

Reply:
I think this looks OK now.